app/server/controller/src/main/java/io/syndesis/server/controller/integration/online/PublishHandler.java

```diff
@@ -238,25 +252,51 @@ private InputStream createProjectFiles(Integration integrationDeployment) {
     // Some helper method to conditional execute certain steps
     @FunctionalInterface
     public interface IoCheckedFunction<T> {
-        String apply(T t, DeploymentData data) throws IOException;
+        String apply(T t, DeploymentData data);
+    }
+
+    @FunctionalInterface
+    public interface IoCheckedErrorHandlingFunction<T> {
+        String apply(T t, DeploymentData data, IntegrationErrorHandler errorHandler);
     }
 
     private class BuildStepOncePerformer {
         private final Map<String, String> stepsPerformed;
         private final IntegrationDeployment integrationDeployment;
+        private IntegrationDeploymentError error;
 
         BuildStepOncePerformer(IntegrationDeployment integrationDeployment) {
             this.integrationDeployment = integrationDeployment;
             this.stepsPerformed = new HashMap<>(integrationDeployment.getStepsDone());
         }
 
-        void perform(String step, IoCheckedFunction<IntegrationDeployment> callable, DeploymentData data) throws IOException {
+        void perform(String step, IoCheckedFunction<IntegrationDeployment> callable, DeploymentData data) {
             if (!stepsPerformed.containsKey(step)) {
-                    stepsPerformed.put(step, callable.apply(integrationDeployment, data));
-                } else {
-                    logInfo(integrationDeployment, "Skipped step {} because already performed", step);
-                }
+                stepsPerformed.put(step, callable.apply(integrationDeployment, data));
+            } else {
+                logInfo(integrationDeployment, "Skipped step {} because already performed", step);
             }
+        }
+
+        void perform(String step, IoCheckedErrorHandlingFunction<IntegrationDeployment> callable, DeploymentData data) {
+            if (!stepsPerformed.containsKey(step)) {
+                stepsPerformed.put(step, callable.apply(integrationDeployment, data,
+                        (throwable) -> error = new IntegrationDeploymentError.Builder()
+                                                    .type(throwable.getClass().getName())
+                                                    .message(throwable.getMessage())
```

Comment:
Perhaps it would be good to log the stack trace of that `throwable`, I know we do it for build step, could be easily forgotten and then difficult to debug.